### PR TITLE
Fix WaitForConnectionAsync when NamedPipeServerStream is disposed

### LINKED ISSUE #40674: Disposing NamedPipeServerStream does not cancel WaitForConnectionAsync
Calling `Dispose` an a `NamedPipeServerStream` which is currently inside `WaitForConnectionAsync` does not cancel the underlying listen event when there are multiple `NamedPipeServerStream` instances active on the same named pipe. This can result in future `NamedPipeClientStream.Connect` calls going to the disposed `NamedPipeServerStream` instead of the non-disposed ones. 

Consider the example below. This will start two `NamedPipeServerStream` instances on a single pipe name, then it will dispose the first one and start a client connection to the pipe. The connection will succeed but it connects to the disposed `NamedPipeServerStream`, not the active one. The stream can then be used without error. 

Example code:

```cs
using System;
using System.IO.Pipes;
using System.Threading;
using System.Threading.Tasks;

class Program
{
    static async Task Main(string[] args)
    {
        var pipeName = Guid.NewGuid().ToString();
        using var server1 = CreateServer();
        using var server2 = CreateServer();

        var wait1 = server1.WaitForConnectionAsync();
        server1.Dispose();

        var waitTask = Task.Run(async () =>
        {
            await server2.WaitForConnectionAsync();
            Console.WriteLine("Server 2 got a connection");
        });

        var client = new NamedPipeClientStream(pipeName);
        await client.ConnectAsync();
        Console.WriteLine("Client connected");
        await wait1;
        Console.WriteLine(wait1.Status);

        // Can still use server1 even though it's disposed
        await server1.WriteAsync(new byte[]{ 42 }, 0, 1);
        var buffer = new byte[1];
        await client.ReadAsync(buffer, 0, 1);
        Console.WriteLine(buffer[0]); // Prints 42
       
        Console.WriteLine("At the end");

        NamedPipeServerStream CreateServer() => new NamedPipeServerStream(
            pipeName,
            PipeDirection.InOut,
            maxNumberOfServerInstances: 10,
            PipeTransmissionMode.Byte,
            PipeOptions.Asynchronous);
    }
}
```

When run on Unix this code will output:

```txt
Client connected
RanToCompletion
42
At the end
```

When run on Windows this code will output:

```txt
Server 2 got a connection
Client connected
Unhandled exception. System.IO.IOException: The pipe has been ended.
   at System.IO.Pipes.ConnectionCompletionSource.HandleError(Int32 errorCode)
   at System.IO.Pipes.PipeCompletionSource`1.CompleteCallback(Int32 resultState)
   at System.IO.Pipes.PipeCompletionSource`1.AsyncCallback(UInt32 errorCode, UInt32 numBytes)   at System.IO.Pipes.ConnectionCompletionSource.AsyncCallback(UInt32 errorCode, UInt32 numBytes)
   at System.IO.Pipes.PipeCompletionSource`1.<>c.<.ctor>b__11_0(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* pOverlapped)
   at System.Threading.ThreadPoolBoundHandleOverlapped.CompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* nativeOverlapped)
   at System.Threading._IOCompletionCallback.PerformIOCompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* pNativeOverlapped)
--- End of stack trace from previous location ---
   at Program.Main(String[] args) in P:\temp\console\Program.cs:line 26
   at Program.<Main>(String[] args)
```

### PR DESCRIPTION
**Pull request by manandre**
Fixes #40674

/cc @jaredpar @davidfowl

### CODE REVIEW

File: src/libraries/System.IO.Pipes/tests/PipeStreamConformanceTests.cs [hunk 1]
```diff
@@ -633,6 +651,26 @@ public async Task CancelTokenOn_Client_ReadWriteCancelledToken_Throws_OperationC
                 await Assert.ThrowsAnyAsync<OperationCanceledException>(() => clientWriteToken);
             }
         }
+
+        [Fact]
+        public async Task TwoServerInstances_OnceDisposed_Throws()
+        {
+            string pipeName = GetUniquePipeName();
+            var server1 = CreateServerStream(pipeName, 2);
```
Comment (stephentoub) [inline src/libraries/System.IO.Pipes/tests/PipeStreamConformanceTests.cs:659]:
```suggestion
            Stream server1 = CreateServerStream(pipeName, 2);
```

File: src/libraries/System.IO.Pipes/tests/PipeStreamConformanceTests.cs [hunk 2]
```diff
@@ -633,6 +651,26 @@ public async Task CancelTokenOn_Client_ReadWriteCancelledToken_Throws_OperationC
...
+            using var server2 = CreateServerStream(pipeName, 2);
+
+            var wait1 = server1.WaitForConnectionAsync();
+            var wait2 = server2.WaitForConnectionAsync();
```
Comment (stephentoub) [inline src/libraries/System.IO.Pipes/tests/PipeStreamConformanceTests.cs:663]:
```suggestion
            Task wait1 = server1.WaitForConnectionAsync();
            Task wait2 = server2.WaitForConnectionAsync();
```

File: src/libraries/System.IO.Pipes/tests/PipeStreamConformanceTests.cs [hunk 3]
```diff
@@ -633,6 +651,26 @@ public async Task CancelTokenOn_Client_ReadWriteCancelledToken_Throws_OperationC
...
+            server1.Dispose();
+
+            using var client = CreateClientStream(pipeName);
```
Comment (stephentoub) [inline src/libraries/System.IO.Pipes/tests/PipeStreamConformanceTests.cs:666]:
```suggestion
            using Stream client = CreateClientStream(pipeName);
```

File: src/libraries/System.IO.Pipes/tests/PipeStreamConformanceTests.cs [hunk 4]
```diff
@@ -633,6 +651,26 @@ public async Task CancelTokenOn_Client_ReadWriteCancelledToken_Throws_OperationC
...
+            await client.ConnectAsync();
+
+            await Assert.ThrowsAsync<IOException>(() => wait1);
+            await ValidateDisposedExceptionsAsync(server1);
+
+            await wait2;
```
Comment (stephentoub) [inline src/libraries/System.IO.Pipes/tests/PipeStreamConformanceTests.cs:672]: We should add validation that the established connection between the client and server2 is usable.

File: src/libraries/System.IO.Pipes/tests/PipeStreamConformanceTests.cs [hunk 5]
```diff
@@ -63,7 +63,16 @@ public abstract class NamedPipeStreamConformanceTests : PipeStreamConformanceTes
...
-        protected abstract (NamedPipeServerStream Server, NamedPipeClientStream Client) CreateServerAndClientStreams();
+        protected abstract NamedPipeServerStream CreateServerStream(string pipeName, int maxInstances = 1);
+        protected abstract NamedPipeClientStream CreateClientStream(string pipeName);
+
+        protected (NamedPipeServerStream Server, NamedPipeClientStream Client) CreateServerAndClientStreams()
+        {
+            string pipeName = GetUniquePipeName();
+            var server = CreateServerStream(pipeName);
+            var client = CreateClientStream(pipeName);
+            return (server, client);
```
Comment (stephentoub) [inline src/libraries/System.IO.Pipes/tests/PipeStreamConformanceTests.cs:74]:
```suggestion
            Stream server = CreateServerStream(pipeName);
            Stream client = CreateClientStream(pipeName);
            return (server, client);
```

File: src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Unix.cs [hunk 6]
```diff
@@ -87,6 +87,12 @@ private void HandleAcceptedSocket(Socket acceptedSocket)
 
             try
             {
+                if (State == PipeState.Closed)
+                {
+                    // Pipe was closed/disposed during the connection phase, handle it as a broken pipe
+                    throw new IOException(SR.IO_PipeBroken);
+                }
```
Comment (stephentoub) [inline src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Unix.cs:94]: Looks reasonable. Thanks.

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.IO.Pipes/tests/PipeStreamConformanceTests.cs (hunk 4 above)
Comment (manandre) [reply src/libraries/System.IO.Pipes/tests/PipeStreamConformanceTests.cs:672]: The test currently hangs on `await wait2` when running on Unix OSes :/
I am still investigating why... but any hint is welcome.

Comment (stephentoub) [reply src/libraries/System.IO.Pipes/tests/PipeStreamConformanceTests.cs:672]: Doesn't that suggest the proposed fix doesn't actually fix the problem? :)

Comment (manandre) [reply src/libraries/System.IO.Pipes/tests/PipeStreamConformanceTests.cs:672]: After further analysis, it seems that this scenario is currently not possible/supported on Unix.
On `dispose` request with multiple servers, the current `accept` operation on `server1` is not (i.e. cannot be) cancelled.
https://github.com/dotnet/runtime/blob/62712ec2913a9f4a1e2e7d512a4d8b0f3299709b/src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Unix.cs#L261-L284
So, the pipe client is connected to `server1` and this connection is now correctly closed/disposed, but it is too late. Then, the `accept` on server2 hangs forever.

I have just discovered that implementation of AcceptAsync cancellation overloads is in progress via #53340. Once available, we could add a cancellation token source to generate a token (linked to user cancellation token) for `AcceptAsync` calls that would be cancelled on `dispose` request.
Does it look reasonable to you?

Comment (geoffkizer) [reply src/libraries/System.IO.Pipes/tests/PipeStreamConformanceTests.cs:672]: Are you saying that with unix pipes, disposing the listen socket does not cancel any outstanding accepts? Or am I misunderstanding?

Comment (manandre) [reply src/libraries/System.IO.Pipes/tests/PipeStreamConformanceTests.cs:672]: No, the ListeningSocket is shared among all server instances targeting the same pipe name. Then, it is not disposed here, as only the first server instance is disposed. So it is why I propose to cancel the accept operation, when the server instance is disposed, via a dedicated cancelation token (source).

Comment (manandre) [reply src/libraries/System.IO.Pipes/tests/PipeStreamConformanceTests.cs:668]: @stephentoub Now the PR does fix the problem ;)
Moreover, I have extended the test to validate that the established connection between the client and server2 is usable.

File: src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Unix.cs [hunk 7]
```diff
@@ -116,9 +138,20 @@ private void HandleAcceptedSocket(Socket acceptedSocket)
...
+        {
             Interlocked.Exchange(ref _instance, null)?.Dispose(disposing); // interlocked to avoid shared state problems from erroneous double/concurrent disposes
 
+            if (disposing)
+            {
+                if (State != PipeState.Closed)
+                {
+                    _internalTokenSource.Cancel();
+                }
+                _internalTokenSource.Dispose();
```
Comment (stephentoub) [inline src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Unix.cs:151]:
```suggestion
```

File: src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Unix.cs [hunk 8]
```diff
@@ -77,8 +78,29 @@ public Task WaitForConnectionAsync(CancellationToken cancellationToken)
...
+                {
+                    //if cancellation was via external token, throw an OCE
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    throw new IOException(SR.IO_PipeBroken);
+                }
+                finally
+                {
+                    linkedTokenSource?.Dispose();
+                }
```
Comment (stephentoub) [inline src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Unix.cs:100]:
```suggestion
                CancellationTokenSource linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(_internalTokenSource.Token, cancellationToken);
                try
                {
                    acceptedSocket = await _instance!.ListeningSocket.AcceptAsync(linkedTokenSource.Token).ConfigureAwait(false);
                }
                catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
                {
                    throw new IOException(SR.IO_PipeBroken);
                }
                finally
                {
                    linkedTokenSource.Dispose();
                }
```

**Review by stephentoub — APPROVED** [review]

**Pull request merged**